File: app/src/main/java/io/github/gustavlindberg99/weather/MainActivity.java
package io.github.gustavlindberg99.weather;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;

import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import com.matthewtamlin.sliding_intro_screen_library.indicators.DotIndicator;

import java.io.IOException;
import java.util.Objects;
import java.util.TimeZone;

public class MainActivity extends AppCompatActivity implements LocationListener{
    private FragmentAdapter _fragmentAdapter;
    private DotIndicator _dotIndicator;
    private boolean _askForLocationPermission = true;

    @SuppressWarnings("deprecation")
    private class FragmentAdapter extends FragmentPagerAdapter{
        private @Nullable WeatherFragment _currentLocationFragment = null;
        private final BiMap<Integer, WeatherFragment> _otherLocationFragments = HashBiMap.create();

        public FragmentAdapter(@NonNull FragmentManager fm){
            super(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
        }

        @Override
        public @NonNull Fragment getItem(int position){
            final @Nullable City currentLocation = City.currentLocation(MainActivity.this);
            if(currentLocation != null){
                position--;
            }
            if(position == -1){
                this._currentLocationFragment = WeatherFragment.getInstance(Objects.requireNonNull(currentLocation));
                return this._currentLocationFragment;
            }
            final WeatherFragment fragment = WeatherFragment.getInstance(City.listFromPreferences(MainActivity.this)[position]);
            this._otherLocationFragments.put(position, fragment);
            return fragment;
        }

        @Override
        public int getItemPosition(@NonNull Object fragment){
            if(!(fragment instanceof WeatherFragment)){
                return POSITION_NONE;
            }

            final boolean currentLocationKnown = City.currentLocation(MainActivity.this) != null;
            if(fragment == this._currentLocationFragment){
                if(currentLocationKnown){
                    return 0;
                }
                else{
                    return POSITION_NONE;
                }
            }

            final Integer position = this._otherLocationFragments.inverse().get(fragment);
            if(position == null){
                return POSITION_NONE;
            }
            return position + (currentLocationKnown ? 1 : 0);
        }

        @Override
        public int getCount(){
            final int currentLocationCount = (City.currentLocation(MainActivity.this) != null) ? 1 : 0;
            final int otherLocationsCount = City.listFromPreferences(MainActivity.this).length;
            return currentLocationCount + otherLocationsCount;
        }

        public @Nullable WeatherFragment currentLocationFragment(){
            return this._currentLocationFragment;
        }

        public WeatherFragment[] allFragments(){
            final int currentLocationFragmentCount = (this._currentLocationFragment != null) ? 1 : 0;
            final int otherLocationFragmentCount = this._otherLocationFragments.size();
            final int fragmentCount = currentLocationFragmentCount + otherLocationFragmentCount;
            WeatherFragment[] fragments = new WeatherFragment[fragmentCount];
            int i = 0;
            if(this._currentLocationFragment != null){
                fragments[i] = this._currentLocationFragment;
                i++;
            }
            for(WeatherFragment fragment: this._otherLocationFragments.values()){
                fragments[i] = fragment;
                i++;
            }
            return fragments;
        }
    }

    public City selectedCity(){
        int selectedIndex = this._dotIndicator.getSelectedItemIndex();
        final @Nullable City currentLocation = City.currentLocation(MainActivity.this);
        if(currentLocation != null){
            selectedIndex--;
        }
        final City[] cities = City.listFromPreferences(this);
        if(selectedIndex >= cities.length){
            selectedIndex = cities.length - 1;
        }
        if(selectedIndex == -1){
            return currentLocation;
        }
        return cities[selectedIndex];
    }

    @Override
    protected void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        this.setContentView(R.layout.activity_main);

        //Make the background of the status bar show correctly
        this.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);

        //Set the fragments
        ViewPager viewPager = this.findViewById(R.id.viewpager);
        this._fragmentAdapter = new FragmentAdapter(this.getSupportFragmentManager());
        viewPager.setAdapter(this._fragmentAdapter);

        //Set the dot indicator
        this._dotIndicator = this.findViewById(R.id.dot_indicator);
        this._dotIndicator.setNumberOfItems(this._fragmentAdapter.getCount());
        this._dotIndicator.setSelectedItem(0, false);
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener(){
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels){}

            @Override
            public void onPageSelected(int position){
                MainActivity.this._dotIndicator.setSelectedItem(position, true);
            }

            @Override
            public void onPageScrollStateChanged(int state){}
        });

        //Initialize the settings button
        final ActivityResultLauncher<Intent> settingsLauncher = this.registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), (ActivityResult result) -> {
            for(WeatherFragment fragment: this._fragmentAdapter.allFragments()){
                fragment.refreshFromCache();
            }
        });
        this.findViewById(R.id.settings_button).setOnClickListener((View v) ->
            settingsLauncher.launch(new Intent(this, Settings.class))
        );

        //Initialize the city list button
        final ActivityResultLauncher<Intent> cityListLauncher = this.registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), (ActivityResult result) -> {
            this._fragmentAdapter.notifyDataSetChanged();
            this._dotIndicator.setNumberOfItems(this._fragmentAdapter.getCount());
            final Intent intent = result.getData();
            if(intent != null){
                final int selectedIndex = intent.getIntExtra(CityList.SELECTED_CITY, viewPager.getCurrentItem());
                this._dotIndicator.setSelectedItem(selectedIndex, false);
                viewPager.setCurrentItem(selectedIndex, false);
            }
        });
        this.findViewById(R.id.city_list_button).setOnClickListener((View v) ->
            cityListLauncher.launch(new Intent(this, CityList.class))
        );

        //Initialize the background
        final @ColorInt int color = this.getColor(R.color.skyBlue);
        ((ImageView) this.findViewById(R.id.background_gradient)).setColorFilter(color);
        ((ImageView) this.findViewById(R.id.background_color)).setColorFilter(color);
    }

    @Override
    protected void onResume(){
        super.onResume();

        //Update the weather at all locations
        for(City city: City.listFromPreferences(this)){
            city.updateWeatherFromServer(city::updateCachedWeather, () -> {});
        }

        //Request a location update
        if(ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED){
            if(this._askForLocationPermission){
                this._askForLocationPermission = false;
                this.requestPermissions(new String[]{Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION}, 0);
            }
            else if(City.listFromPreferences(this).length == 0){
                this.findViewById(R.id.city_list_button).callOnClick();
            }
        }
        else{
            ((LocationManager) this.getSystemService(Context.LOCATION_SERVICE)).requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, this);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults){
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if(ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED){
            if(City.listFromPreferences(this).length == 0){
                this.findViewById(R.id.city_list_button).callOnClick();
            }
        }
        else{
            ((LocationManager) this.getSystemService(Context.LOCATION_SERVICE)).requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, this);
        }
    }

    @Override
    public void onLocationChanged(Location location){
        final boolean previousLocationKnown = (City.currentLocation(this) != null);
        new Thread(() -> {
            try{
                final Address address = new Geocoder(this).getFromLocation(location.getLatitude(), location.getLongitude(), 1).get(0);
                City city = new City(this, AddressFunctions.cityNameFromAddress(this, address), location.getLatitude(), location.getLongitude(), TimeZone.getDefault().getID());

                city.updateWeatherFromServer(() -> {
                    city.saveAsCurrentLocation();

                    final WeatherFragment currentLocationFragment = this._fragmentAdapter.currentLocationFragment();
                    if(currentLocationFragment != null){
                        currentLocationFragment.setLocation(city);
                    }
                    if(!previousLocationKnown){
                        this._fragmentAdapter.notifyDataSetChanged();
                        this._dotIndicator.setNumberOfItems(this._fragmentAdapter.getCount());
                    }
                }, () -> {
                    //If finding the weather at the new location failed, just keep the previous location
                });
            }
            catch(IOException ignore){
                //If finding the location failed, just keep the previous location
            }
        }).start();
    }
}

File: app/src/main/java/io/github/gustavlindberg99/weather/MainActivity.kt
@file:Suppress("DEPRECATION")

package io.github.gustavlindberg99.weather

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Geocoder
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.ColorInt
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import com.matthewtamlin.sliding_intro_screen_library.indicators.DotIndicator
import java.io.IOException
import java.util.*

class MainActivity: AppCompatActivity(), LocationListener {
    private val _fragmentAdapter: FragmentAdapter by lazy{FragmentAdapter(this.supportFragmentManager)}
    private val _dotIndicator: DotIndicator by lazy{this.findViewById(R.id.dot_indicator)}
    private var _askForLocationPermission: Boolean = true

    private inner class FragmentAdapter(fm: FragmentManager): FragmentPagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT){
        private var _currentLocationFragment: WeatherFragment? = null
        private val _otherLocationFragments: BiMap<Int, WeatherFragment> = HashBiMap.create()
        override fun getItem(position: Int): Fragment {
            val currentLocation = City.currentLocation(this@MainActivity)
            val truePosition = if(currentLocation != null) position - 1 else position
            if(truePosition == -1){
                val currentLocationFragment = WeatherFragment.getInstance(currentLocation!!)    //currentLocation can't be null here otherwise truePosition wouldn't be negative and this block won't be entered, see how truePosition is defined above
                this@FragmentAdapter._currentLocationFragment = currentLocationFragment
                return currentLocationFragment
            }
            val fragment = WeatherFragment.getInstance(City.listFromPreferences(this@MainActivity)[truePosition])
            this@FragmentAdapter._otherLocationFragments[truePosition] = fragment
            return fragment
        }

        override fun getItemPosition(fragment: Any): Int {
            if(fragment !is WeatherFragment){
                return POSITION_NONE
            }
            val currentLocationKnown = City.currentLocation(this@MainActivity) != null
            if(fragment === this@FragmentAdapter._currentLocationFragment){
                return if(currentLocationKnown) 0 else POSITION_NONE
            }
            val position = this@FragmentAdapter._otherLocationFragments.inverse()[fragment] ?: return POSITION_NONE
            return position + if(currentLocationKnown) 1 else 0
        }

        override fun getCount(): Int {
            val currentLocationCount = if(City.currentLocation(this@MainActivity) != null) 1 else 0
            val otherLocationsCount = City.listFromPreferences(this@MainActivity).size
            return currentLocationCount + otherLocationsCount
        }

        fun currentLocationFragment(): WeatherFragment? {
            return this@FragmentAdapter._currentLocationFragment
        }

        fun allFragments(): List<WeatherFragment> {
            val fragments: MutableList<WeatherFragment> = mutableListOf()
            val currentLocationFragment = this@FragmentAdapter._currentLocationFragment
            if(currentLocationFragment != null){
                fragments.add(currentLocationFragment)
            }
            for(fragment in this@FragmentAdapter._otherLocationFragments.values){
                fragments.add(fragment)
            }
            return fragments
        }
    }

    override fun onCreate(savedInstanceState: Bundle?){
        super.onCreate(savedInstanceState)
        this.setContentView(R.layout.activity_main)

        //Make the background of the status bar show correctly
        this.window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN

        //Set the fragments
        val viewPager: ViewPager = this.findViewById(R.id.viewpager)
        viewPager.adapter = this._fragmentAdapter

        //Set the dot indicator
        this._dotIndicator.numberOfItems = this._fragmentAdapter.count
        this._dotIndicator.setSelectedItem(0, false)
        viewPager.addOnPageChangeListener(object: OnPageChangeListener{
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int){}
            override fun onPageSelected(position: Int){
                this@MainActivity._dotIndicator.setSelectedItem(position, true)
            }
            override fun onPageScrollStateChanged(state: Int){}
        })

        //Initialize the settings button
        val settingsLauncher = this.registerForActivityResult(ActivityResultContracts.StartActivityForResult()){
            for(fragment in this._fragmentAdapter.allFragments()){
                fragment.refreshFromCache()
            }
        }
        this.findViewById<View>(R.id.settings_button).setOnClickListener{settingsLauncher.launch(Intent(this, Settings::class.java))}

        //Initialize the city list button
        val cityListLauncher = this.registerForActivityResult(ActivityResultContracts.StartActivityForResult()){result: ActivityResult ->
            this._fragmentAdapter.notifyDataSetChanged()
            this._dotIndicator.numberOfItems = this._fragmentAdapter.count
            val intent = result.data
            if(intent != null){
                val selectedIndex = intent.getIntExtra(CityList.SELECTED_CITY, viewPager.currentItem)
                this._dotIndicator.setSelectedItem(selectedIndex, false)
                viewPager.setCurrentItem(selectedIndex, false)
            }
        }
        this.findViewById<View>(R.id.city_list_button).setOnClickListener{cityListLauncher.launch(Intent(this, CityList::class.java))}

        //Initialize the background
        @ColorInt val color = getColor(R.color.skyBlue)
        this.findViewById<ImageView>(R.id.background_gradient).setColorFilter(color)
        this.findViewById<ImageView>(R.id.background_color).setColorFilter(color)
    }

    override fun onResume(){
        super.onResume()

        //Update the weather at all locations
        for(city in City.listFromPreferences(this)){
            city.updateWeatherFromServer({city.updateCachedWeather()}, {})
        }

        //Request a location update
        if(ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED){
            if(this._askForLocationPermission){
                this._askForLocationPermission = false
                this.requestPermissions(arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION), 0)
            }
            else if(City.listFromPreferences(this).isEmpty()){
                this.findViewById<View>(R.id.city_list_button).callOnClick()
            }
        }
        else{
            //LocationManager.NETWORK_PROVIDER is much faster than LocationManager.GPS_PROVIDER
            (this.getSystemService(LOCATION_SERVICE) as LocationManager).requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 60000, 100f, this)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray){
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if(ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED){
            if(City.listFromPreferences(this).isEmpty()){
                this.findViewById<View>(R.id.city_list_button).callOnClick()
            }
        }
        else {
            (this.getSystemService(LOCATION_SERVICE) as LocationManager).requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 60000, 100f, this)
        }
    }

    override fun onLocationChanged(location: Location){
        val previousLocationKnown = City.currentLocation(this) != null
        Thread{
            try{
                val address = Geocoder(this).getFromLocation(location.latitude, location.longitude, 1)!![0]
                val city = City(this, address.cityName(this), location.latitude, location.longitude, TimeZone.getDefault().id)
                city.updateWeatherFromServer({
                    city.saveAsCurrentLocation()
                    val currentLocationFragment = this._fragmentAdapter.currentLocationFragment()
                    currentLocationFragment?.setLocation(city)
                    if(!previousLocationKnown){
                        this._fragmentAdapter.notifyDataSetChanged()
                        this._dotIndicator.numberOfItems = this._fragmentAdapter.count
                    }
                }, {})
            }
            catch(ignore: IOException){
                //If finding the location failed, just keep the previous location
            }
        }.start()
    }

    //These are abstract on some versions of Android but not others, so they must be implemented otherwise it will give a runtime error, see https://stackoverflow.com/a/67900719/4284627
    override fun onProviderEnabled(provider: String){}
    override fun onProviderDisabled(provider: String){}
    @Deprecated("Deprecated in Java")
    override fun onStatusChanged(provider: String, status: Int, extras: Bundle){}

    fun selectedCity(): City? {
        var selectedIndex: Int = this._dotIndicator.selectedItemIndex
        val currentLocation: City? = City.currentLocation(this)
        if(currentLocation != null){
            selectedIndex--
        }
        val cities: List<City> = City.listFromPreferences(this)
        if(selectedIndex >= cities.size){
            selectedIndex = cities.size - 1
        }
        if(selectedIndex == -1){
            return currentLocation
        }
        return cities[selectedIndex]
    }
}